python/pyarrow/dataset.py

```diff
@@ -310,23 +303,23 @@ def dataset(sources, filesystem=None, partitioning=None, format=None):
 
     >>> dataset("path/to/nyc-taxi/", format="parquet")
 
-    Combining different sources:
+    Construction from multiple factories:
 
     >>> dataset([
-    ...     source("s3://old-taxi-data", format="parquet"),
-    ...     source("local/path/to/new/data", format="csv")
+    ...     factory("s3://old-taxi-data", format="parquet"),
+    ...     factory("local/path/to/new/data", format="csv")
     ... ])
 
     """
-    if not isinstance(sources, list):
-        sources = [sources]
+    if not isinstance(paths_or_factories, list):
+        paths_or_factories = [paths_or_factories]
 
-    sources = [
-        _ensure_source(src, filesystem=filesystem, partitioning=partitioning,
+    factories = [
+        _ensure_factory(src, filesystem=filesystem, partitioning=partitioning,
                        format=format)
-        for src in sources
+        for src in paths_or_factories
     ]
-    return DatasetFactory(sources).finish()
+    return TreeDatasetFactory(factories).finish()
```

Comment:
Should we special case the "single FileSystemDataset child" case, so when doing `dataset("my/dir/")` gives a FileSystemDataset instead of TreeDataset?

Reply:
Sure, that makes sense.